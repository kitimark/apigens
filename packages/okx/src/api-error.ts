import { AxiosError, AxiosResponse, InternalAxiosRequestConfig } from 'axios';

export class OkxError<T = unknown, D = any> extends AxiosError<T, D> {
  constructor(
    message?: string,
    code?: string,
    config?: InternalAxiosRequestConfig<D>,
    request?: any,
    response?: AxiosResponse<T, D>,
  ) {
    super(message, code, config, request, response);
    this.name = 'OkxError';
  }
}
