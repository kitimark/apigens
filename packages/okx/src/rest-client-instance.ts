import axios, { AxiosInstance } from 'axios';
import { InternalAxiosRequestConfig } from 'axios';
import * as querystring from 'querystring';
import { createHmac } from 'crypto';
import { AxiosResponse } from 'axios';
import { OkxResponse } from './api-core-types';
import { OkxError } from './api-error';

export interface OkxConfig {
  demo?: boolean;
  url?: string;  
  apiKey: string;
  secretKey: string;
  passphrase: string;
}

export type OkxAxiosInstance = AxiosInstance;

export const createOkxAxiosInstance = (config: OkxConfig): OkxAxiosInstance => {
  const client = axios.create({
    baseURL: config.url ?? 'https://www.okx.com/',
  })

  /**
   * Sign request for call okx api service
   * ref: https://www.okx.com/docs-v5/en/#rest-api-authentication
   *
   * @param request
   * @returns request with credential headers
   */
  const generateCredentials = (request: InternalAxiosRequestConfig<any>) => {
    /**
     * Create signature for call okx api service
     * ref: https://www.okx.com/docs-v5/en/#rest-api-authentication-signature
     * @param request
     * @returns Signature and timestamp
     */
    const createSignature = (request: InternalAxiosRequestConfig<any>) => {
      const date = new Date();
      const isoDate = date.toISOString();

      let data = isoDate + request.method!.toUpperCase() + request.url;
      if (request.params) {
        data = data + '?' + querystring.stringify(request.params);
      }
      if (request.data) {
        data = data + JSON.stringify(request.data);
      }
      const hmac = createHmac('sha256', config.secretKey);
      hmac.update(data);

      return {
        timestamp: isoDate,
        signature: hmac.digest('base64'),
      };
    };
    const { timestamp, signature } = createSignature(request);

    request.headers.set('OK-ACCESS-KEY', config.apiKey);
    request.headers.set('OK-ACCESS-PASSPHRASE', config.passphrase);
    request.headers.set('OK-ACCESS-SIGN', signature);
    request.headers.set('OK-ACCESS-TIMESTAMP', timestamp);

    /**
     * Check env for return is this demo account.
     * ref: https://www.okx.com/docs-v5/en/#overview-demo-trading-services
     * @returns return true, if okx demo account
     */
    if (config.demo) {
      request.headers.set('x-simulated-trading', 1);
    }

    return request;
  };
  client.interceptors.request.use(generateCredentials);

  /**
   * Error handler for okx api service
   * ref: https://www.okx.com/docs-v5/en/#error-code
   */
  const errorHandler = (response: AxiosResponse<OkxResponse<any>>) => {
    if (!['0', '2'].includes(response.data.code)) {
      throw new OkxError(
        response.data.msg,
        response.data.code,
        response.config,
        response.request,
        response,
      );
    }
    return response;
  }
  client.defaults.validateStatus = (status) => {
    return status < 500;
  }
  client.interceptors.response.use(errorHandler);

  return client;
}
