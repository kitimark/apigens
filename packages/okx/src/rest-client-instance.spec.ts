import { AxiosResponse } from 'axios';
import { OkxAxiosInstance, createOkxAxiosInstance } from './rest-client-instance';
import { OkxResponse } from './api-core-types';
import { OkxError } from './api-error';

describe('rest-client-instance', () => {
  let api: OkxAxiosInstance;
  let mockAdapter = jest.fn();

  beforeAll(() => {
    api = createOkxAxiosInstance({
      apiKey: 'test-api-key',
      secretKey: 'test-secret-key',
      passphrase: 'test-passpharse',
    });

    api.defaults.adapter = mockAdapter;
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  describe('signature', () => {
    test('request get', async () => {
      jest.useFakeTimers({
        now: new Date(0),
      });
      mockAdapter.mockResolvedValue(<AxiosResponse<OkxResponse>>{
        data: {
          code: '0',
          msg: '',
          data: <any>[],
        },
      });

      await api.get('/test-api');

      expect(mockAdapter).toHaveBeenCalledTimes(1);
      expect(mockAdapter.mock.lastCall[0].headers).toMatchSnapshot();
    });

    test('request get with querystring', async () => {
      jest.useFakeTimers({ now: 0 });
      mockAdapter.mockResolvedValue(<AxiosResponse<OkxResponse>>{
        data: {
          code: '0',
          msg: '',
          data: <any>[],
        },
      });

      await api.get('/test-api', {
        params: {
          param1: 'value',
        },
      });

      expect(mockAdapter).toHaveBeenCalledTimes(1);
      expect(mockAdapter.mock.lastCall[0].headers).toMatchSnapshot();
    });

    test('request post', async () => {
      jest.useFakeTimers({ now: 0 });
      mockAdapter.mockResolvedValue(<AxiosResponse<OkxResponse>>{
        data: {
          code: '0',
          msg: '',
          data: <any>[],
        },
      });

      await api.post('/test-api', {
        param1: 'value'
      });

      expect(mockAdapter).toHaveBeenCalledTimes(1);
      expect(mockAdapter.mock.lastCall[0].headers).toMatchSnapshot();
    });
  });

  describe('error handler', () => {
    test.each(['0', '2'])('succes with %s code', async (code) => {
      mockAdapter.mockResolvedValue(<AxiosResponse<OkxResponse>>{
        data: {
          code,
          msg: '',
          data: <any>[]
        }
      });

      await api.get('/');

      expect(mockAdapter).toHaveBeenCalledTimes(1);
    });

    test.each(['1', '50000', '50001'])('error with %s code', async (code) => {
      mockAdapter.mockResolvedValue(<AxiosResponse<OkxResponse>>{
        data: {
          code,
          msg: '',
          data: <any>[]
        }
      });

      expect(api.get('/')).rejects.toThrowError(OkxError);
    })
  });
});
