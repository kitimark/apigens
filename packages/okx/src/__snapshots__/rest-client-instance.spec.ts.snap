// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rest-client-instance signature request get 1`] = `
{
  "Accept": "application/json, text/plain, */*",
  "OK-ACCESS-KEY": "test-api-key",
  "OK-ACCESS-PASSPHRASE": "test-passpharse",
  "OK-ACCESS-SIGN": "EQTtq7fY3jChCxmJ/kQ6vg+ieuMkjA9scJeis6/lU18=",
  "OK-ACCESS-TIMESTAMP": "1970-01-01T00:00:00.000Z",
}
`;

exports[`rest-client-instance signature request get with querystring 1`] = `
{
  "Accept": "application/json, text/plain, */*",
  "OK-ACCESS-KEY": "test-api-key",
  "OK-ACCESS-PASSPHRASE": "test-passpharse",
  "OK-ACCESS-SIGN": "PGDrFQWozEO+oyGvdvhds8u0a8Otl6gxGSKrsGSbyag=",
  "OK-ACCESS-TIMESTAMP": "1970-01-01T00:00:00.000Z",
}
`;

exports[`rest-client-instance signature request post 1`] = `
{
  "Accept": "application/json, text/plain, */*",
  "Content-Type": "application/json",
  "OK-ACCESS-KEY": "test-api-key",
  "OK-ACCESS-PASSPHRASE": "test-passpharse",
  "OK-ACCESS-SIGN": "XzYey6LLdxDdYVRXLuqchBjsLE2SLD5rCWqJXetHpD0=",
  "OK-ACCESS-TIMESTAMP": "1970-01-01T00:00:00.000Z",
}
`;
